### DISCUSSION TITLE
WordPress architecture from foundation to skyscrapers

### DISCUSSION THREAD

**Discussion by oglekler** [WordPress/developer-blog-content#111] [Topic Ideas]
## Proposal

Series of topics about software architecture with regard to WordPress and practical examples of implementation in themes, plugins and the core itself.

### Why this is important

Assumption is that architectural concepts can be easily overlooks and aren't obvious and usual methods of learning things even in dedicated courses are leaving gaps in this area. And if the basement of the project is unstable, the whole build on top of it can be shaking violently and turn into a nightmare for users and developer itself.

## Topic for series:

- WordPress architecture — How to use WordPress core in its full extent
- CSS for WordPress — How to stop messing with styles
- JS for WordPress — Easy maintainable scripts without dirty hacks
- PHP for WordPress — PHP projects as your portfolio / Writing PHP you will be thanked for
- Creating WordPress theme and not have a headache
- Creating WordPress plugin which will last

## The Goal
The goal is to show readers a way to figure out what project structure they need and why, and not present ready to use solutions as the only available options or the truth in the flesh. Still, proven methods needs to be presented fully with explanation why they are good. When you know the reasoning behind, it is easy to understand if something will fulfil your needs or not.

### Additional goals can be:

- present existing architectural solutions with examples
- direct readers to existing best practised documentation
- show how to keep in mind scalability, test covering, maintainability and SOLID-DRY-KISS principles
- functional and object-oriented programming (this is always the topic for holy wars)
- EDA vs MVC
- how to reduce errors
- mention used conventions or their absence
- make clear difference between architectural approaches best fitted to WP with other well known solutions
- hot to remove memory leaks, reduce CPU and memory usage, remove unnecessary request
- how to reduce dependencies, especially from third party applications
- how to make projects with high load in mind
- etc.

This is just a proposal, ideas are still quite vague and I am not insisting on anything 🙂

**Comment by marybaum** [comment]
It occurs to me that Tonya would be a great resource for this. I was a member of her Know the Code course from 2016 on, and one of its objectives was to get WordPressers thinking more like computer scientists.... from my phoneOn May 6, 2023, at 6:52 AM, Olga Gleckler ***@***.***> wrote:﻿
Proposal
Series of topics about software architecture with regard to WordPress and practical examples of implementation in themes, plugins and the core itself.
Why this is important
Assumption is that architectural concepts can be easily overlooks and aren't obvious and usual methods of learning things even in dedicated courses are leaving gaps in this area.
Topic for series:

WordPress architecture — How to use WordPress core in its full extent
CSS for WordPress — How to stop messing with styles
JS for WordPress — Easy maintainable scripts without dirty hacks
PHP for WordPress — PHP projects as your portfolio / Writing PHP you will be thanked for
Creating WordPress theme and not have a headache
Creating WordPress plugin which will last

The Goal
The goal is to show readers a way to figure out what project structure they need and why, and not present ready to use solutions as the only available options or the truth in the flesh. Still, proven methods needs to be presented fully with explanation why they are good. When you know the reasoning behind, it is easy to understand if something will fulfil your needs or not.
Additional goals can be:

present existing architectural solutions with examples
direct readers to existing best practised documentation
show how to keep in mind scalability, test covering, maintainability and SOLID-DRY-KISS principles
functional and object-oriented programming (this is always the topic for holy wars)
EDA vs MVC
how to reduce errors
mention used conventions or their absence
make clear difference between architectural approaches best fitted to WP with other well known solutions
hot to remove memory leaks, reduce CPU and memory usage, remove unnecessary request
how to reduce dependencies, especially from third party applications
how to make project with high load in mind
etc.

This is just a proposal, ideas are still quite vague and I am not insisting on anything 🙂

**Comment by bph** [comment]
[from the meeting notes:](https://make.wordpress.org/core/2023/06/02/developer-blog-editorial-meeting-01-june-2023/) "With regards to the second topic that was discussed, it was considered that the remit is too broad but that there may be some good ideas for individual posts contained within it."
